File: SpearoGo/Views/TidesPage.swift
import SwiftUI

struct TidesPage: View {
    // Matched to the high/low columns, which are taller now the type is bigger.
    @ScaledMetric(relativeTo: .title3) private var dividerHeight: CGFloat = 52

    @Environment(AppState.self) private var appState
    @State private var crownOffset: Double = 0

    var body: some View {
        ScrollView {
            VStack(spacing: Brand.Spacing.section) {
                Text("Tides")
                    .brandSectionHeader()

                if let tide = appState.tideData {
                    HStack(spacing: 16) {
                        TideEventView(label: "HIGH", event: tide.nextHigh(), tide: tide)

                        Divider()
                            .frame(height: dividerHeight)
                            .background(Brand.Colors.textSecondary.opacity(Brand.Opacity.borderLine))

                        TideEventView(label: "LOW", event: tide.nextLow(), tide: tide)
                    }
                    .accessibilityElement(children: .combine)
                    .accessibilityLabel(tideTimesLabel(tide))

                    TideDirectionView(isRising: tide.isRising(), phase: tide.phase())
                        .brandCard(padding: Brand.Spacing.item)
                        .accessibilityLabel("Tide is \(tide.isRising() ? "incoming" : "outgoing"), phase: \(tide.phase().rawValue)")

                    // Current height
                    HStack(spacing: Brand.Spacing.micro) {
                        Text("Now")
                            .itemLabelStyle()
                        Text(tide.currentHeight().map { String(format: "%.1fm", $0) } ?? "—")
                            .highlightCaptionStyle()
                    }
                    .accessibilityLabel(tide.currentHeight().map { String(format: "Current tide height %.1f metres", $0) } ?? "Current tide height not reported")

                    if let source = sourceLabel(tide) {
                        Text(source)
                            .captionStyle()
                            .lineLimit(1)
                    }
                } else if appState.isLoading {
                    // Shimmer skeleton
                    HStack(spacing: 16) {
                        TideEventSkeleton()
                        SkeletonBlock(width: 1, height: 44)
                        TideEventSkeleton()
                    }
                    SkeletonBlock(width: 120, height: 30)
                    .accessibilityLabel("Loading tide data")
                } else {
                    // Not a loading state. The synthetic tide model this page
                    // used to draw was wrong everywhere on earth, so it was
                    // removed rather than improved. Real predictions arrive
                    // with the tidesGo backend.
                    Text("No tide data for this spot yet")
                        .captionStyle()
                        .multilineTextAlignment(.center)
                        .accessibilityLabel("No tide data for this spot yet")
                }
            }
            .padding(Brand.Spacing.page)
        }
        .focusable()
        .digitalCrownRotation($crownOffset)
        .brandPage()
    }

    private func tideTimesLabel(_ tide: TideData) -> String {
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        func describe(_ name: String, _ event: TideEvent?) -> String {
            guard let event else { return "No further \(name) tide today" }
            return "\(name) tide at \(TideEventView.stationTime.string(from: tide.stationLocalDate(event))), "
                + "\(String(format: "%.1f", event.height)) metres"
        }
        return "\(describe("High", tide.nextHigh())). \(describe("Low", tide.nextLow()))"
    }

    /// Where the numbers came from. A named gauge and an ocean-model estimate
    /// must never read as equally trustworthy: the model is materially weaker
    /// in the estuaries and inlets these users dive.
    private func sourceLabel(_ tide: TideData) -> String? {
        if tide.isStale { return "Saved forecast" }
        if tide.isModelEstimate { return "Estimated · no nearby gauge" }
        return tide.stationName
    }
}

struct TideEventView: View {
    /// UTC, because the value handed to it is already shifted into the
    /// station's local time.
    static let stationTime: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm"
        f.timeZone = TimeZone(secondsFromGMT: 0)
        return f
    }()

    let label: String
    /// nil after the day's last turn — shown as "—" rather than wrapped round
    /// to a time that has already passed.
    let event: TideEvent?
    /// Carries the station's UTC offset, so the time shown is the one the local
    /// tide table prints rather than the reader's wall clock.
    let tide: TideData

    var body: some View {
        VStack(spacing: Brand.Spacing.micro) {
            Text(label).itemLabelStyle()
            if let event {
                Text(Self.stationTime.string(from: tide.stationLocalDate(event)))
                    .timeDisplayStyle()
                Text(String(format: "%.1fm", event.height)).highlightCaptionStyle()
            } else {
                Text("—").timeDisplayStyle()
                Text(" ").highlightCaptionStyle()
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(event.map { "\(label) tide: \(String(format: "%.1f", $0.height)) metres" }
                            ?? "No further \(label.lowercased()) tide today")
    }
}

struct TideEventSkeleton: View {
    var body: some View {
        VStack(spacing: Brand.Spacing.micro) {
            SkeletonBlock(width: 30, height: 8)
            SkeletonBlock(width: 50, height: 16)
            SkeletonBlock(width: 35, height: 9)
        }
    }
}

struct TideDirectionView: View {
    let isRising: Bool
    let phase: TidePhase

    var body: some View {
        HStack(spacing: Brand.Spacing.item) {
            Image(systemName: isRising ? "arrow.up.circle.fill" : "arrow.down.circle.fill")
                .foregroundStyle(isRising ? Brand.Colors.maybe : Brand.Colors.primary)
                .font(.system(size: 18))

            VStack(alignment: .leading, spacing: 0) {
                Text(isRising ? "Incoming" : "Outgoing")
                    .brandFont(Brand.Typography.personalityCopy)
                    .foregroundStyle(Brand.Colors.textPrimary)
                Text(phase.rawValue)
                    .captionStyle()
            }
        }
    }
}

// MARK: - Previews

#Preview {
    TidesPage()
        .previewAsWatch()
        .environment(AppState.preview())
}

#Preview("Loading") {
    TidesPage()
        .previewAsWatch()
        .environment(AppState.previewLoading())
}

File: wear/app/src/main/java/com/spearotracker/spearogo/ui/pages/TidesPage.kt
package com.spearotracker.spearogo.ui.pages

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.wear.compose.material3.ScreenScaffold
import androidx.wear.compose.material3.Text
import com.spearotracker.spearogo.models.TidePhase
import com.spearotracker.spearogo.ui.AppUiState
import com.spearotracker.spearogo.ui.theme.Brand
import java.text.SimpleDateFormat
import java.util.*
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.style.TextAlign
import java.util.TimeZone

@Composable
fun TidesPage(uiState: AppUiState) {
    val scrollState = rememberScrollState()

    ScreenScaffold(scrollState = scrollState) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(scrollState)
            .padding(Brand.Spacing.page),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Text(
            text = "TIDES",
            style = Brand.Typography.sectionHeader,
            color = Brand.Colors.textSecondary,
            modifier = Modifier.padding(bottom = Brand.Spacing.item)
        )

        val tide = uiState.tideData

        if (tide != null) {
            // UTC, because the value handed to it is already shifted into the
            // station's local time. Using the device zone here is what showed a
            // Lagos 23:50 low as 00:50 on a German wrist.
            val timeFormat = SimpleDateFormat("HH:mm", Locale.getDefault()).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }

            Row(
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                modifier = Modifier.padding(bottom = Brand.Spacing.section)
            ) {
                // HIGH
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(text = "HIGH", style = Brand.Typography.itemLabel, color = Brand.Colors.textSecondary)
                    Text(
                        text = tide.nextHigh()?.let { timeFormat.format(Date(tide.stationLocalMillis(it))) } ?: "—",
                        style = Brand.Typography.timeDisplay,
                        color = Brand.Colors.textPrimary
                    )
                    Text(
                        text = tide.nextHigh()?.let { "%.1fm".format(it.height) } ?: "",
                        style = Brand.Typography.caption,
                        color = Brand.Colors.secondary
                    )
                }

                // Divider
                // Matched to the high/low columns, which are taller now the
                // type is bigger.
                val dividerHeight = (52 * LocalDensity.current.fontScale).dp
                Box(
                    modifier = Modifier
                        .width(1.dp)
                        .height(dividerHeight)
                        .background(Brand.Colors.textSecondary.copy(alpha = Brand.Opacity.borderLine))
                )

                // LOW
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(text = "LOW", style = Brand.Typography.itemLabel, color = Brand.Colors.textSecondary)
                    Text(
                        text = tide.nextLow()?.let { timeFormat.format(Date(tide.stationLocalMillis(it))) } ?: "—",
                        style = Brand.Typography.timeDisplay,
                        color = Brand.Colors.textPrimary
                    )
                    Text(
                        text = tide.nextLow()?.let { "%.1fm".format(it.height) } ?: "",
                        style = Brand.Typography.caption,
                        color = Brand.Colors.secondary
                    )
                }
            }

            // Tide direction card
            Row(
                modifier = Modifier
                    .background(
                        Brand.Colors.textPrimary.copy(alpha = Brand.Opacity.cardFill),
                        RoundedCornerShape(Brand.Radius.card)
                    )
                    .border(
                        1.dp,
                        Brand.Colors.textPrimary.copy(alpha = Brand.Opacity.borderLine),
                        RoundedCornerShape(Brand.Radius.card)
                    )
                    .padding(Brand.Spacing.item),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Brand.Spacing.item)
            ) {
                Text(
                    text = if (tide.isRising()) "\u2191" else "\u2193",
                    style = Brand.Typography.dataValue,
                    color = if (tide.isRising()) Brand.Colors.maybe else Brand.Colors.primary
                )
                Column {
                    Text(
                        text = if (tide.isRising()) "Incoming" else "Outgoing",
                        style = Brand.Typography.personalityCopy,
                        color = Brand.Colors.textPrimary
                    )
                    Text(
                        text = tide.phase().label,
                        style = Brand.Typography.caption,
                        color = Brand.Colors.textSecondary
                    )
                }
            }

            Spacer(modifier = Modifier.height(Brand.Spacing.item))

            // Current height. Null when the day's hourly readings do not
            // bracket this moment — shown as "—" rather than guessed.
            Row(horizontalArrangement = Arrangement.spacedBy(Brand.Spacing.micro)) {
                Text(text = "Now", style = Brand.Typography.itemLabel, color = Brand.Colors.textSecondary)
                Text(
                    text = tide.currentHeight()?.let { "%.1fm".format(it) } ?: "—",
                    style = Brand.Typography.caption,
                    color = Brand.Colors.secondary
                )
            }

            // Where the numbers came from. A named gauge and an ocean-model
            // estimate must never read as equally trustworthy: the model is
            // materially weaker in the estuaries and inlets these users dive.
            val source = when {
                tide.isStale -> "Saved forecast"
                tide.isModelEstimate -> "Estimated \u00b7 no nearby gauge"
                tide.stationName != null -> tide.stationName
                else -> null
            }
            source?.let {
                Spacer(modifier = Modifier.height(Brand.Spacing.micro))
                Text(
                    text = it,
                    style = Brand.Typography.caption,
                    color = Brand.Colors.textSecondary,
                    textAlign = TextAlign.Center,
                    maxLines = 1
                )
            }
        } else if (uiState.isLoading) {
            Text(
                text = "Loading tides...",
                style = Brand.Typography.caption,
                color = Brand.Colors.textSecondary
            )
        } else {
            // Not a loading state. The synthetic tide model this page used to
            // draw was wrong everywhere on earth, so it was removed rather than
            // improved. Real predictions arrive with the tidesGo backend.
            Text(
                text = "No tide data for this spot yet",
                style = Brand.Typography.caption,
                color = Brand.Colors.textSecondary,
                textAlign = TextAlign.Center
            )
        }
    }
    }
}
